aip/general/0193.md

```diff
@@ -75,132 +86,191 @@ following:
+ `google.rpc.Status` already exists and can’t be changed. In this
+ case, do not edit anything in this `google.rpc.Status`, because some
+ users might have automation running against the pre-existing
+ content.  For example, any changes in text for the `message` string
+ could cause breakages. For more information, see [Changing Error
+ Messages](#changing-error-messages).
 
-**Note:** `ErrorInfo` represents a special case. There **must** be exactly one
-`ErrorInfo`. It is required.
 
-### Error messages
+##### [google.rpc.Status payload][Status]
 
-For each error, the service **must** populate the `message` field on
-[`google.rpc.Status`][Status]. This error message,
+Details of this object are summarized in the following fields, field
+descriptions, and examples:
 
-- is a developer-facing, human-readable "debug message"
-- is presented in the service's native language
-- both explains the error and offers an actionable resolution to it
-  ([citation](https://cloud.google.com/apis/design/errors#error_model))
 
-**Note:** Sometimes a service will elect to always present
-`Status.message` in English rather than the application's native
-language so that messages are easily searchable in common knowledge
-bases, such as StackOverflow&trade;.
+- `code int32`
+  - The status code, which **should** be an enum value of
+    [`google.rpc.Code`][Code].
+  - *Example*: 429
+- `message string`
+  - A developer-facing error message, which **should** be in English.
+    Any user-facing error message **should** be located in the
+    LocalizedMessage payload.
+  - *Example message*: \"The zone \'us-east1-a\' does not have enough
+    resources available to fulfill the request. Try a different zone, or
+    try again later.\"
+- `details Object (repeated)`
+  - Text
+  - *Example*: Described in the [ErrorDetails](#error-details) section.
 
-When introducing an error that represents a failure scenario that did
-not previously occur for the service, the payload **must** include
-`ErrorInfo` and any variables found in dynamic segments of the error
-message **must** be present in `ErrorInfo.metadata`. See, "[Dynamic
-variables](#dynamic-variables)".
 
-#### Changing error messages
+#### [ErrorDetails][details]
 
-<a name="status-message-warning"></a>
+Google defines a set of [standard detail payloads][details] for error
+details, which cover most common needs for API errors.  Services
+**should** use these standard detail payloads when feasible.
 
-`Status.message` **may** change. However, **use extreme caution** when
-doing so.  API consumers often treat this error message as **part of the
-API contract**. Clients perform string matches on the text to
-differentiate one error for another and even parse the error message to
-extract variables from dynamic segments.
+Structured details with machine-readable identifiers **must** be used if
+users will need to write code against a specific aspect of the error.
```

Comment:
This makes it sound like variables in message segments only need to be included in ErrorInfo.metadata if the API author thinks that their API will be used to write code against. How will they know? This language **must** be much firmer. Any variables in messages **must** be in  the metadata.

Reply:
For what it's worth, I believe this was existing content of the AIP. That said, I agree that the language is confusing.

@jschneider11 What about changing this to "Structured details with machine-readable identifiers must be used so that users can write code against specific aspects of the error."?

Reply:
SGTM. Thanks.